amespahdbpythonsuite/transitions.py

```diff
@@ -1,10 +1,544 @@
+        used when multiprocessing is required.
+
+        Parameters
+        ----------
+
+        uid : int
+            single UID value.
+
+        Returns
+        -------
+        temp : dict
+            Dictionary of Tmax.
+        ud : dict
+            Dictionary of calculated intensities for the given UID.
+
+        """
+        global energy
+        energy = e
+
+        global frequencies
+        frequencies = np.array([d['frequency'] for d in self.data[uid]])
+
+        global intensities
+        intensities = np.array([d['intensity'] for d in self.data[uid]])
+
+        Tmax = optimize.brentq(self.attainedtemperature, 2.73, 5000.0)
+
+        for d in self.data[uid]:
+            if d['intensity'] > 0:
+                global frequency
+                frequency = d['frequency']
+                d['intensity'] *= d['frequency'] ** 3 * integrate.quad(self.featurestrength, 2.73, Tmax)[0]
+
+        temp = {'uid': uid, 'temperature': Tmax}
+        ud = {uid: self.data[uid]}
+        return ud, temp
+
+    def cascade(self, e, **keywords):
+        """
+        Applies the Cascade emission model.
+
+        Parameters
+        ----------
+        e : float
+            Excitation energy in erg.
+
+        """
+        if self.type != 'theoretical':
+            message('THEORETICAL DATABASE REQUIRED FOR EMISSION MODEL')
+            return
+
+        if self.model:
+            if self.model['type'] != 'zerokelvin_m':
+                message(f'AN EMISSION MODEL HAS ALREADY BEEN APPLIED: {self.model["type"]}')
+                return
+
+        message('APPLYING CASCADE EMISSION MODEL')
+
+        tstart = time.perf_counter()
+
+        global energy
+
+        energy = e
+
+        self.model = {'type': 'cascade_m',
+                      'energy': e,
+                      'temperatures': [],
+                      'description': ''}
+
+        self.units['ordinate'] = {'unit': 3,
+                                  'str': 'integrated radiant energy [erg]'}
+
+        print(57 * '=')
+
+        if keywords.get('multiprocessing'):
+            cascade_em_model = partial(self._cascade_em_model, e)
+            if keywords.get('ncores'):
+                ncores = keywords.get('ncores')
+                print(f'Using multiprocessing module with {ncores} cores')
+                pool = multiprocessing.Pool(processes=ncores)
+                data, temp = zip(*pool.map(cascade_em_model, self.uids))
+            else:
+                print('Using multiprocessing module with 2 cores')
+                pool = multiprocessing.Pool(processes=2)
+                data, temp = zip(*pool.map(cascade_em_model, self.uids))
+            pool.close()
+            pool.join()
+
+            # Re-assign self.data.
+            self.data = {}
+            for d in list(data):
+                for k, v in d.items():
+                    self.data[k] = v
+
+                self.model['temperatures'] = temp
```

Comment:
Could you verify with me that the indentation here is redundant? Should be 1 tab back. Probably is moved there by accident. If so, could we move it in the main code? Otherwise I can fix it in the next PR.